File: packages/js/test/client.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { CrawlFox, CrawlFoxError } from "../src/index.js";

describe("CrawlFox client", () => {
  it("requires an API key", () => {
    const prev = process.env.CRAWLFOX_API_KEY;
    delete process.env.CRAWLFOX_API_KEY;
    assert.throws(() => new CrawlFox(), /API key required/);
    process.env.CRAWLFOX_API_KEY = prev;
  });

  it("scrape sends correct request and returns a Document", async () => {
    const calls: { url: string; init: RequestInit }[] = [];
    const fetchMock = async (url: string | URL | Request, init?: RequestInit) => {
      calls.push({ url: String(url), init: init ?? {} });
      return new Response(
        JSON.stringify({
          success: true,
          data: { markdown: "# Hello", metadata: { statusCode: 200 } },
        }),
        { status: 200, headers: { "content-type": "application/json" } },
      );
    };

    const client = new CrawlFox({ apiKey: "cfx_test", fetch: fetchMock });
    const res = await client.scrape("https://example.com", {
      formats: ["markdown"],
      skipCache: true,
      country: "de",
      language: "de",
      redactPII: true,
    });

    assert.equal(res.success, true);
    assert.equal(res.markdown, "# Hello");
    assert.equal(calls.length, 1);
    assert.match(calls[0].url, /\/v1\/scrape$/);
    const headers = calls[0].init.headers as Record<string, string>;
    assert.equal(headers.Authorization, "Bearer cfx_test");
    assert.match(headers["User-Agent"], /^crawlfox-js\//);
    const body = JSON.parse(String(calls[0].init.body));
    assert.equal(body.url, "https://example.com");
    assert.deepEqual(body.formats, ["markdown"]);
    assert.equal(body.skipCache, true);
    assert.equal(body.country, "de");
    assert.equal(body.redactPII, true);
  });

  it("search sends query body", async () => {
    let body: Record<string, unknown> = {};
    const fetchMock = async (_url: string | URL | Request, init?: RequestInit) => {
      body = JSON.parse(String(init?.body));
      return new Response(
        JSON.stringify({
          success: true,
          data: { web: [{ url: "https://example.com", title: "Ex" }] },
        }),
        { status: 200, headers: { "content-type": "application/json" } },
      );
    };

    const client = new CrawlFox({ apiKey: "cfx_test", fetch: fetchMock });
    const res = await client.search("crawlfox", { engine: "google", num: 5 });
    assert.equal(res.web?.[0]?.url, "https://example.com");
    assert.equal(body.q, "crawlfox");
    assert.equal(body.engine, "google");
    assert.equal(body.num, 5);
  });

  it("batch returns documents under data", async () => {
    const fetchMock = async () =>
      new Response(
        JSON.stringify({
          success: true,
          count: 1,
          results: [{ success: true, data: { markdown: "# A" } }],
        }),
        { status: 200, headers: { "content-type": "application/json" } },
      );
    const client = new CrawlFox({ apiKey: "cfx_test", fetch: fetchMock });
    const batch = await client.batch(["https://example.org"], { formats: ["markdown"] });
    assert.equal(batch.data[0].markdown, "# A");
  });

  it("retries retryable errors then succeeds", async () => {
    let n = 0;
    const fetchMock = async () => {
      n += 1;
      if (n === 1) {
        return new Response(
          JSON.stringify({
            code: "RATE_LIMIT_EXCEEDED",
            message: "slow down",
            retryable: true,
          }),
          { status: 429, headers: { "content-type": "application/json" } },
        );
      }
      return new Response(
        JSON.stringify({ success: true, data: { markdown: "# ok" } }),
        { status: 200, headers: { "content-type": "application/json" } },
      );
    };
    const client = new CrawlFox({ apiKey: "cfx_test", fetch: fetchMock, maxRetries: 2 });
    const res = await client.scrape("https://example.com");
    assert.equal(res.markdown, "# ok");
    assert.equal(n, 2);
  });

  it("throws CrawlFoxError on non-retryable failure", async () => {
    const fetchMock = async () =>
      new Response(
        JSON.stringify({
          code: "MONTHLY_QUOTA_EXCEEDED",
          message: "Quota exceeded",
          retryable: false,
        }),
        { status: 429, headers: { "content-type": "application/json" } },
      );

    const client = new CrawlFox({ apiKey: "cfx_test", fetch: fetchMock });
    await assert.rejects(
      () => client.scrape("https://example.com"),
      (err: unknown) => {
        assert.ok(err instanceof CrawlFoxError);
        assert.equal(err.status, 429);
        assert.equal(err.code, "MONTHLY_QUOTA_EXCEEDED");
        return true;
      },
    );
  });

  it("searchStream yields NDJSON events", async () => {
    const ndjson = [
      JSON.stringify({ type: "page", data: { web: [{ url: "https://a.com" }] } }),
      JSON.stringify({ type: "done", partial: false, creditsUsed: 1, id: "x" }),
      "",
    ].join("\n");
    const fetchMock = async () =>
      new Response(ndjson, {
        status: 200,
        headers: { "content-type": "application/x-ndjson" },
      });
    const client = new CrawlFox({ apiKey: "cfx_test", fetch: fetchMock });
    const events = [];
    for await (const ev of client.searchStream("q", { num: 20 })) {
      events.push(ev);
    }
    assert.equal(events.length, 2);
    assert.equal(events[0].type, "page");
    assert.equal(events[1].type, "done");
  });

  it("getLog hits /v1/logs/:id", async () => {
    let url = "";
    const fetchMock = async (u: string | URL | Request) => {
      url = String(u);
      return new Response(
        JSON.stringify({
          id: "abc",
          started_at_ms: 1,
          duration_ms: 2,
          url: "https://example.com",
          status: 200,
        }),
        { status: 200, headers: { "content-type": "application/json" } },
      );
    };
    const client = new CrawlFox({ apiKey: "cfx_test", fetch: fetchMock });
    const row = await client.getLog("abc");
    assert.match(url, /\/v1\/logs\/abc$/);
    assert.equal(row.status, 200);
  });
});

File: packages/ruby/test/client_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "crawlfox"

class ClientTest < Minitest::Test
  def test_requires_api_key
    prev = ENV.delete("CRAWLFOX_API_KEY")
    assert_raises(ArgumentError) { Crawlfox::Client.new }
  ensure
    ENV["CRAWLFOX_API_KEY"] = prev if prev
  end

  def test_scrape_search_batch_logs_and_retry
    scrapes = 0
    transport = lambda do |method, url, headers, body|
      case url
      when %r{/v1/scrape$}
        scrapes += 1
        return [503, { message: "busy", retryable: true }.to_json] if scrapes == 1

        assert_equal "POST", method
        assert_equal "Bearer cfx_test", headers["Authorization"]
        assert headers["User-Agent"].start_with?("crawlfox-ruby/")
        payload = JSON.parse(body)
        assert_equal "https://example.com", payload["url"]
        [200, { success: true, data: { markdown: "# Hello" } }.to_json]
      when %r{/v1/search$}
        [200, { success: true, data: { web: [{ url: "https://example.com" }] } }.to_json]
      when %r{/v1/batch$}
        [200, { success: true, count: 1, results: [{ success: true, data: { markdown: "# A" } }] }.to_json]
      when %r{/v1/logs/log1/result$}
        [200, { ok: true }.to_json]
      when %r{/v1/logs/log1$}
        assert_equal "GET", method
        [200, { id: "log1", status: "ok" }.to_json]
      else
        [404, "{}"]
      end
    end

    app = Crawlfox::Client.new(api_key: "cfx_test", api_url: "https://api.test", retry_backoff_ms: 1, transport: transport)
    doc = app.scrape("https://example.com", formats: ["markdown"], country: "de")
    assert_equal "# Hello", doc["markdown"]
    assert_equal 2, scrapes

    search = app.search("crawlfox", engine: "google", num: 5)
    assert_equal "https://example.com", search["web"][0]["url"]

    batch = app.batch(["https://example.org"])
    assert_equal "# A", batch["data"][0]["markdown"]

    assert_equal "ok", app.get_log("log1")["status"]
    assert_equal true, app.get_log_result("log1")["ok"]
  end

  def test_does_not_retry_quota
    n = 0
    transport = lambda do |*|
      n += 1
      [429, { message: "quota", retryable: false, code: "quota" }.to_json]
    end
    app = Crawlfox::Client.new(api_key: "cfx_test", api_url: "https://api.test", retry_backoff_ms: 1, transport: transport)
    err = assert_raises(Crawlfox::Error) { app.scrape("https://example.com") }
    assert_equal 429, err.status
    assert_equal 1, n
  end
end

class LiveTest < Minitest::Test
  def test_live_scrape_search_batch
    skip "CRAWLFOX_API_KEY not set" if ENV["CRAWLFOX_API_KEY"].to_s.empty?

    app = Crawlfox::Client.new(max_retries: 1)
    doc = app.scrape("https://example.com", formats: ["markdown"])
    assert doc["markdown"].to_s.length > 10
    search = app.search("crawlfox", engine: "google", num: 3)
    refute_empty search["web"]
    batch = app.batch(["https://example.com"], formats: ["markdown"])
    assert batch.dig("data", 0, "markdown").to_s.length > 5
  end
end
